lib/web/app/auth.go

```diff
@@ -18,45 +18,137 @@ package app
+//	    that makes the cookie unaccesible and hard to tamper with. We use this
+//	    "double submit cookie" method to protect the entire auth exchange flow
+//	    from CSRF.
+//
+//	2): If the "state" query param is present, we will serve a blank HTML page
+//	    that has inline JS that contains logic to complete the auth exchange.
+func (h *Handler) startAppAuthExchange(w http.ResponseWriter, r *http.Request, p httprouter.Params) error {
+	q := r.URL.Query()
+
+	// Initiate auth exchange.
+	if q.Get("state") == "" {
+		// secretToken is the token we will look for in both the cookie
+		// and in the request "state" query param.
+		secretToken, err := utils.CryptoRandomHex(auth.TokenLenBytes)
+		if err != nil {
+			h.log.WithError(err).Debugf("Failed to generate and encode random numbers.")
+			return trace.AccessDenied("access denied")
+		}
+
+		// cookieIdentifier is used to uniquely identify this cookie
+		// that will be used to store this secret token.
+		//
+		// This prevents a race condition (state token mismatch error)
+		// where we can overwrite existing cookie (with the same name) with a
+		// different token value eg: launch app in multiple tabs in quick succession
+		cookieIdentifier, err := utils.CryptoRandomHex(auth.TokenLenBytes)
+		if err != nil {
+			h.log.WithError(err).Debugf("Failed to generate and encode random numbers.")
+			return trace.AccessDenied("access denied")
+		}
+
+		h.setAuthStateCookie(w, secretToken, cookieIdentifier)
+
+		webLauncherURLParams := launcherURLParams{
+			clusterName: q.Get("cluster"),
+			publicAddr:  q.Get("addr"),
+			arn:         q.Get("arn"),
+			path:        q.Get("path"),
+			// The state token concats both the secret token and the cookie ID.
+			// The server will break this token to its individual parts:
+			//   - secretToken to compare against the one stored in cookie
+			//   - cookieIdentifier to look up cookie sent by browser.
+			stateToken: fmt.Sprintf("%s_%s", secretToken, cookieIdentifier),
+		}
+		return h.redirectToLauncher(w, r, webLauncherURLParams)
+	}
+
+	// Continue the auth exchange.
+
+	nonce, err := utils.CryptoRandomHex(auth.TokenLenBytes)
+	if err != nil {
+		h.log.WithError(err).Debugf("Failed to generate and encode random numbers.")
+		return trace.AccessDenied("access denied")
+	}
+	SetRedirectPageHeaders(w.Header(), nonce)
+
+	// Serving the HTML page.
+	fmt.Fprintf(w, appRedirectionJs, nonce)
+	return nil
+}
+
+// completeAppAuthExchange completes the auth exchange flow started by "startAppAuthExchange" handler
+// by validating the values passed in the request body, and upon success sets cookies required
+// for the current app session. User should be able to interact with the app now.
+func (h *Handler) completeAppAuthExchange(w http.ResponseWriter, r *http.Request, p httprouter.Params) error {
 	httplib.SetNoCacheHeaders(w.Header())
+	var req fragmentRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return trace.Wrap(err)
```

Comment:
Since the primary reason this would fail is invalid JSON, what do you think about `trace.BadParameter` instead?

Or `trace.AccessDenied`, which is what you use below.